### ISSUE TITLE
OWLObjectInverseOf - argument type is too general 

### ISSUE THREAD

**Issue by sesuncedu** [owlcs/owlapi#430]
We have 
```java
public interface OWLObjectInverseOf extends OWLObjectPropertyExpression {
    @Nonnull
    OWLObjectPropertyExpression getInverse();
}
```

However, the spec says:
```
ObjectProperty := IRI
ObjectPropertyExpression := ObjectProperty | InverseObjectProperty
InverseObjectProperty := 'ObjectInverseOf' '(' ObjectProperty ')'
```

So the method should really return OWLObjectProperty, and the corresponding factory method and implementation constructors should only accept OWLObjectProperty.

**Comment by ignazio1977** [comment]
But the mapping to RDF has both inverseOf(OP) and inverseOf(OPE), where OP is object property and OPE is object property expression http://www.w3.org/TR/owl2-mapping-to-rdf/

For a change I'm confused by these corner cases in the specs.

**Comment by sesuncedu** [comment]
All quotes are from the RDF mapping document (supra).

Notation
-----------
Notation is introduced in §§ 1 and 2.1 
> *:x denotes an IRI;   _:x denotes a blank node;
>OP denotes an object property; OPE denotes an object property expression;

Structural Syntax to RDF:
------------------------

Table 1 defines the transformation from structural syntax to triples. The four relevant lines are excerpted. 

>**Table 1. Transformation to Triples.**

>Element E of the Structural Specification | Triples Generated in an Invocation of T(E)	 | Main Node of T(E)
------------------------------------------------------|-----------------------------------------------------------|----------------------
OP | | ```OP```
ObjectInverseOf( OP ) | ```_:x owl:inverseOf T(OP) .``` | ```_:x```
ObjectPropertyAssertion( OP a1 a2 )| ```T(a1) T(OP) T(a2) .``` | 
ObjectPropertyAssertion( ObjectInverseOf( OP ) a1 a2 ) | ```T(a2) T(OP) T(a1) .``` | 

From line 2, we know that the main node of an application of ObjectInverseOf is a blank node. 
From line 3, we know that an OP must be an IRI, since the result of T(OP) appears in the predicate position of non-generalized RDF. 
From line 4, we confirm that the argument OP to ObjectInverseOf must be an IRI, since the result of T(OP) is used in the predicate position of a non-generalized RDF. 

RDF to Structural Syntax
---------------------------------
The RDF to Structural Syntax mapping defines a function, OPE, that maps from  a blank node or IRI to an Object Property  Expression

> **§3.1.2 Parsing of the Ontology Header and Declarations**
>Table 7. Parsing Declarations in G

>If G contains this pattern... | ...then this declaration is added to Decl(G).
----|-----
```*:x rdf:type owl:ObjectProperty .``` | Declaration( ObjectProperty( *:x ) )
```_:x rdf:type owl:Axiom . _:x owl:annotatedSource *:y .  _:x owl:annotatedProperty rdf:type . _:x owl:annotatedTarget owl:ObjectProperty . ``` | Declaration( ObjectProperty( *:x ) )

No other patterns generate Declaration(ObjectProperty(x)).  
Thus we confirm that all parsed declarations of ObjectProperty must be for IRIs.

OPE(x) is initialized to empty, then initially populated using the declarations from the imports closure

>Functio[n...] OPE [... is] initialized as shown in Table 9.
>**Table 9. Initialization of CE, DR, OPE, DPE, and AP**

>If AllDecl(G) contains this declaration... | ...then perform this assignment.
----------------------------------------------------|--------------------------------------------
Declaration( ObjectProperty( *:x ) )	| OPE(*:x) := an object property with the IRI *:x

This only defines values of OPE for IRI arguments ; the result is  always a named object property. 
There is no other place defining values of OPE for IRIs. 

The only other source of mappings for OPE(x) is in Table 11.  

> **§3.2.4  Parsing of Expressions.**
> **Table 11. Parsing Object Property Expressions** 

>If G contains this pattern... | ...then OPE(_:x) is set to this object property expression.
-----|-----
 ```_:x owl:inverseOf *:y .``` { OPE(_:x) = ε and OPE(*:y) ≠ ε } |  ObjectInverseOf( OPE(*:y) )

This only defines values of OPE for blank node arguments.  The results are always the result of applying the ObjectInverseOf constructor to the result of applying OPE to an IRI valued argument. 
As we saw above, this must always be a named object property. 

Thus, the RDF mapping document confirms the structural syntax specification and the FSS grammar.

**Comment by sesuncedu** [comment]
Have put in a check in the constructor and data factory, which triggered a cascade of failing unit tests that were checking for the ability to correctly reduce object properties to their simplest form (which involved creating what were structurally illegal forms in order to fix them). 

The only unit test currently failing is DataUnionOfTestCase2, which I suspect is due to illegal input. 

The grammar in OWLFunctionalSyntaxParser.jj is wrong too ; it's a one word fix, if I can remember how to invoke javacc :)

**Comment by ignazio1977** [comment]
@matthewhorridge what are your thoughts? I can't recall if inverses had any other reason to be defined as they are.

Technically this is also an interface change. I'm OK with it, given that it's small, but I don't know the impact. Code working with inverses would be impacted.

**Comment by matthewhorridge** [comment]
My initial thought... I'm against breaking this.

I can't remember what the historical reasons for this were, but I seem to remember it being allowed at some point along the road to standardisation.  It would be better if the API matched the functional syntax spec, but I don't feel like this is a huge issue (in this case) - it seems harmless.

I don't really have strong feelings, so I'd get over it if there was a change.  However, I like backwards compatibility :)

**Comment by sesuncedu** [comment]
The current set of changes treat the error as semantic rather than
syntactic; an exception will be thrown at run time rather than when at
compile time.

If inverse properties are created by calling the inverseProperty method on
a PE then the result will automatically be in simplified form.  If someone
deliberately creates an ontology with multiple inversion, then they should
be made redundant :-P

The advantage of the standard is that inverseProperty becomes a single bit
field on OPE, which makes it easier to handle in a packed binary encoding.

**Comment by ignazio1977** [comment]
Well, it's a change in behaviour rather than a change in interface so code should still compile; might spit out a few more exceptions than before. 
One workaround would be, in the data factory, to intercept multiple wraps and simplify them on the spot. People would lose the ability to double and quadruple wrap inverses syntactically but retain an equivalent result, so little harm done there. Sounds like a backward compatible plan.

This suggests a lint test too: detect needless wrapping and amend it.

I'll merge and amend as suggested.

**Comment by ignazio1977** [comment]
There is no extra fix to be applied without interface changes - an expression cannot be double wrapped in the current code.

I've checked the Protege code base and all the plugins I have checked out locally, including HermiT and Pellet. All uses of the data factory to create OWLObjectInverseOf are called with explicit OWLObjectProperty, minus a couple of occurrences where the code is relying on whatever is passed as input - these look like convenience methods, and they look like they will be called with actual OWLObjectProperty objects anyway. So, I'm fairly confident the code as is will not throw exceptions.

I'll amend the functional syntax parsing.
